### PR TITLE
BUG: groupby().any() returns true for groups with timedelta all NaT

### LINKED ISSUE #59712: BUG: groupby().any() returns True instead of False for groups where timedelta column is all null
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

pd.DataFrame([pd.Timedelta(1), pd.NaT]).groupby([0, 1]).any()
```

### Issue Description

For other dtypes, like integers and strings, groupby().any() returns `True` for groups where all the values are null, e.g.

```python
pd.DataFrame([1, None]).groupby([0, 1]).any()

pd.DataFrame(["a", None]).groupby([0, 1]).any()
```

### Expected Behavior

groupby().any() should return False for groups where all the timedelta values are null. 

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : d9cdd2ee5a58015ef6f4d15c7226110c9aab8140
python                : 3.9.18.final.0
python-bits           : 64
OS                    : Darwin
OS-release            : 23.6.0
Version               : Darwin Kernel Version 23.6.0: Mon Jul 29 21:13:04 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6020
machine               : arm64
processor             : arm
byteorder             : little
LC_ALL                : None
LANG                  : en_US.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.2.2
numpy                 : 1.26.3
pytz                  : 2023.3.post1
dateutil              : 2.8.2
setuptools            : 68.2.2
pip                   : 23.3.1
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : 8.18.1
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2023.4
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by Petroncini**
PR Summary:

This PR addresses an issue where groupby().any() returns True for groups where all Timedelta values are NaT, which is inconsistent with other dtypes (like integers and strings). For non-Timedelta dtypes, groupby().any() correctly returns False when all values in a group are null.
Reproduction:

python

import pandas as pd

# Current behavior: incorrect output for all NaT groups
df = pd.DataFrame([pd.Timedelta(1), pd.NaT]).groupby([0, 1]).any()

For other dtypes (integers and strings):

python

df = pd.DataFrame([1, None]).groupby([0, 1]).any()  # Returns False
df = pd.DataFrame(["a", None]).groupby([0, 1]).any()  # Returns False

Expected behavior for Timedelta values: groupby().any() should return False for groups where all values are NaT.
Fix Details:

    Modified the _call_cython_op function in the groupby aggregation to correctly handle Timedelta types. Now, when all values in a group are NaT, the any() method will return False.
    This ensures consistent behavior across all dtypes, including Timedelta.

Tests Added:

    A new test has been added to check that groupby().any() returns False for groups with only NaT Timedelta values and True for groups with at least one non-NaT Timedelta.

Closing:

This PR fixes the inconsistent behavior of groupby().any() for Timedelta values and improves the overall correctness of the method. Let me know if there are any additional changes needed.

Resolves: #59712

Note: I am a complete novice to open source and this is my first ever PR. I'm taking an open source systems class is college and we're suppposed to attempt to contribute to a existing project we use and I chose pandas as it is a tool that has helped me alot. So i ask for any guidance anyone can give me about any potential mistakes or improvements i can make here. Appreciate it!

### CODE REVIEW

**Comment by mroeschke** [comment]
Thanks for the pull request but this is already being worked on in https://github.com/pandas-dev/pandas/pull/59750. Closing to let that contributor finish the effort, but we can reopen this if it goes stale

**Comment by Petroncini** [comment]
Hi @mroeschke , could this be reopened?

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -374,6 +374,13 @@ def _call_cython_op(
         if is_datetimelike:
             values = values.view("int64")
             is_numeric = True
+
+            # Fix for NaT handling: ensure NaT is treated as False in any() and all()
+            if self.how in ["any", "all"]:
+                # Set NaT (which is represented as the smallest int64) to False (0)
+                nat_mask = values == np.iinfo(np.int64).min
+                values[nat_mask] = 0  # Treat NaT as False
```
Comment (rhshadrach) [inline pandas/core/groupby/ops.py]: The linked issue suggests moving up the determination of `mask` below to be prior to viewing the values as `int64` on L375. Is there a reason this doesn't work?

Comment (Petroncini) [reply pandas/core/groupby/ops.py]: Sorry, it appears i went in the wrong direction with my implementation. Just did a commit with this better solution. Could you give me some pointers on all these check failures?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Merging main should fix the CI issues. However, I do not believe the changes here will fix the issue and will break other tests. We need to move the determination of `mask = isna(values)` to happen before `values = values.view("int64")` but not change the order of any other logic.

File: pandas/tests/groupby/test_grouping.py [hunk 2]
```diff
@@ -1180,3 +1180,18 @@ def test_grouping_by_key_is_in_axis():
     result = gb.sum()
     expected = DataFrame({"a": [1, 2], "b": [1, 2], "c": [7, 5]})
     tm.assert_frame_equal(result, expected)
+
+
+def test_groupby_any_with_timedelta():
+    # Create a DataFrame with a single column containing a Timedelta and NaT
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_grouping.py:1186]: The comments in this test repeat the code verbatim. Can you remove them.

In addition, can you add a reference to the issue as the first line of this test.

```python
def test_groupby_any_with_timedelta():
    # GH#59712
```

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Looking good! This needs a line in the whatsnew. Can you add a line to `docs/source/whatsnew/v3.0.0.rst` in the groupby subsection of the bug fixes.

**Comment by Petroncini** [comment]
Hi @rhshadrach, how's this? I was having some issues with the linux 32 bit tests but it seems resolved now.

**Comment by Petroncini** [comment]
Hi @rhshadrach, how's this?

File: pandas/tests/groupby/test_grouping.py (hunk 2 above)
Comment (Petroncini) [reply pandas/tests/groupby/test_grouping.py]: I added these changes.

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @Petroncini!

**Comment by Petroncini** [comment]
Thanks for all the help and guidance, @rhshadrach!

**Pull request merged**